feat: add jstree json generator to core

Kad1n opened the pull request:
Closes #57 

Allows the UI to be ran on another machine, core is the one who holds sends the file structure.

Kad1n committed:
feat: add jsTree json generator to core

Kad1n committed:
fix: add missing folderB directory

R0flcopt3r commented on src/core/services.py:
```diff
@@ -508,3 +511,73 @@ def queue_job(project_id: int, job_id: int, session: Session) -> None:
+    root_node: Dict[str, Any] = dict()
+    child_list = []
+
+    if isfile(path):
```
Is this necessary when you are checking that you're working on a dir below?

Kad1n replied:
Yes, the one bellow checks if the path is valid. This one checks if it is a file, and not a directory. `os.walk(path)` throws an error if this a file.

R0flcopt3r replied:
`isdir` simply returns `True` if `path` is an existing directory, which means it should also be valid for `os.walk`.

Kad1n replied:
```python
if isfile(path): # Returns FALSE if it is an invalid path (it is not a file)
    raise NotADirectoryError("Path must be a directory.")

if not isdir(path): # If it is NOT a directory (invalid path)
    raise FileNotFoundError("Directory at '{}' was not found.".format(path))
````

Kad1n replied:
Having just one check `isdir` wouldnt make it possible to differentiate the type of error. That way it would throw "file not found" even if the path is valid, but pointing to a file.

R0flcopt3r commented on src/core/services.py:
```diff
@@ -508,3 +511,73 @@ def queue_job(project_id: int, job_id: int, session: Session) -> None:
+    child_list.extend(
+        [
+            {
+                "text": name,
```
Wouldn't this make a little more sense?

```suggestion
                "name": name,
```

MindTooth replied:
Ref.: https://www.jstree.com/docs/json/

MindTooth commented:
I believe the current output is not completely valid.  Look below for a working example of data.

Take note of `children: true`.  This is needed so jsTree renders the node as closed but when opened it sends a new request using said nodes `id`.

---

`http 'localhost:5000/projects/files/'`
```json
[
  {
    "children": [
      {
        "children": true, 
        "id": 2, 
        "text": "Child node 1"
      }, 
      {
        "id": 3, 
        "text": "Child node 2"
      }
    ], 
    "id": 1, 
    "text": "Root node 1", 
    "type": "root"
  }
]
```

`http 'localhost:5000/projects/files/?id=2'`
```json
[
  {
    "children": [
      {
        "id": 4, 
        "text": "Child node 3"
      }, 
      {
        "id": 5, 
        "text": "Child node 4"
      }
    ], 
    "id": 2, 
    "text": "Child node 2"
  }
]
```

---



Currently, I'm using `?id=#` for sending requests.  But I plan to look into making it work using `/#` instead.

Kad1n committed:
feat: add storage endpoint in core api

MindTooth commented on src/core/services.py:
```diff
@@ -508,3 +511,73 @@ def queue_job(project_id: int, job_id: int, session: Session) -> None:
+    # Create root node
+    root_node = {
+        "text": root,
+        "type": "folder",
```
```suggestion
        "type": "root",
```

Could maybe return a list of root nodes.  If multiple folders are set to be mounted.  Internal folder, external drive, etc.

Kad1n replied:
Not sure if this is a good idea, if you request a second time in a subfolder, the selected folder will now be a "root"

MindTooth replied:
I would serve this when no parameter is provided.  Then it will always be true for root.

Kad1n replied:
Ok, that can be arranged. Will work on this tomorrow.

Kad1n committed:
feat: add children boolean tag to subfolders

MindTooth commented:
I've concluded after some testing that we need some sort of `id`.  This can be the path itself.  jsTree needs a way to link back new nodes pulled via lazy loading.  The autogenerated `id` doesn't provide this.

I can continue on this to make it work if that's okay?  Good job so far.

Kad1n commented:
> I've concluded after some testing that we need some sort of `id`. This can be the path itself. jsTree needs a way to link back new nodes pulled via lazy loading. The autogenerated `id` doesn't provide this.
> 
> I can continue on this to make it work if that's okay? Good job so far.

Go ahead and implement it as you want! :smile:

MindTooth commented:
Did you not look into using https://docs.python.org/3/library/mimetypes.html#mimetypes.guess_type?  Noticed that you had hardcoded the types in the test.

<blockquote><img src="../_static/py.png" width="48" align="right"><div><strong><a href="https://docs.python.org/3/library/mimetypes.html#mimetypes.guess_type">mimetypes — Map filenames to MIME types — Python 3.9.6 documentation</a></strong></div></blockquote>

MindTooth committed:
refactor(core): structure for jstree using lazy loading

MindTooth committed:
feat(core): add usage of base64

tomrtk commented on src/core/services.py:
```diff
@@ -575,10 +588,22 @@ def get_directory_listing(path: str) -> Dict[str, Any]:
     )
 
     # Create root node
+    # TODO: Improve naming when having `/` at the end of the query.
```
Could maybe use `os.path.join` on line 580.

Example:
```py
>>> import os
>>> os.path.join("/m/d", "a")
'/m/d/a'
>>> os.path.join("/m/d/", "a")
'/m/d/a'
```

R0flcopt3r commented on src/core/api/api.py:
```diff
@@ -651,3 +656,81 @@ def get_objects_from_job(
+    HTTPException
+        If the path is invalid.
+    """
+    path = pPath(sys.executable).anchor
```
Probably better to use the full namespace, than renaming it. 

 ```py
import pathlib


path = pathlib.Path(…)
```

MindTooth commented:
> Did you not look into using https://docs.python.org/3/library/mimetypes.html#mimetypes.guess_type? Noticed that you had hardcoded the types in the test.
> 
> > <img alt="" width="48" src="../_static/py.png">**[mimetypes — Map filenames to MIME types — Python 3.9.6 documentation](https://docs.python.org/3/library/mimetypes.html#mimetypes.guess_type)**

I saw that you did.  My bad!  Good job on the pull request. 👍🏻 

<blockquote><img src="../_static/py.png" width="48" align="right"><div><strong><a href="https://docs.python.org/3/library/mimetypes.html#mimetypes.guess_type">mimetypes — Map filenames to MIME types — Python 3.9.6 documentation</a></strong></div></blockquote>

MindTooth committed:
test(core): check more

MindTooth commented:
`PermissionError: [Errno 13] Permission denied: '/Library/Application Support/ApplePushService'`

```
  File "/Users/mt/School/core-jstree/src/core/services.py", line 629, in <listcomp>
    "children": len(os.listdir(os.path.join(root, name))) > 0,
```

Handle permission errors.

tomrtk committed:
refactor: test using `pathlib` for generate folder/file id's

  Test using `pathlib` for generating id's in tree structure to se if this
  fix test issues on windows.

tomrtk commented:
Fixed generation of file/folder `id` in `get_directory_listing` by using `pathlib` for cross-platform testing in commit 5d36eadecaf.
Also using `pathlib` to traverse files and folders in `path`.

tomrtk commented on src/core/api/api.py:
```diff
@@ -584,3 +589,81 @@ def get_objects_from_job(
+    HTTPException
+        If the path is invalid.
+    """
+    # TODO: Test this
```
Believe this is tested now?
```suggestion
```

tomrtk commented on src/core/api/api.py:
```diff
@@ -584,3 +589,81 @@ def get_objects_from_job(
+    """
+    # TODO: Test this
+    path64d = base64.urlsafe_b64decode(path)
+    pathd = path64d.decode("utf-8")
```
```suggestion
    pathd = base64.urlsafe_b64decode(path).decode("utf-8")
```

tomrtk commented on src/core/services.py:
```diff
@@ -561,3 +566,92 @@ def get_job_objects(
+    #   "id": "/Applications/////Xcode.app",
+    #   "text": "Xcode.app",
+    #   "type": "folder"
+    # },
```
```suggestion
```

tomrtk commented on src/core/services.py:
```diff
@@ -561,3 +566,92 @@ def get_job_objects(
+    }
+    from pprint import pprint
+
+    pprint(root_node)
```
```suggestion
```

MindTooth commented:
The Windows tests passed, so nice job!

MindTooth commented on src/core/api/api.py:
```diff
@@ -584,3 +589,80 @@ def get_objects_from_job(
+    HTTPException
+        If the path is invalid.
+    """
+    path = pathlib.Path(sys.executable).anchor
```
Make an attempt at setting this using the new config feature of #73.

We've looked at using the startup event from FastAPI/Starlette.  But have been unable to override a global root_path.

Ref.: https://fastapi.tiangolo.com/advanced/testing-events/

```diff
diff --git a/src/core/api/api.py b/src/core/api/api.py
index bf1b5db..daed219 100644
--- a/src/core/api/api.py
+++ b/src/core/api/api.py
@@ -10,7 +10,7 @@ import base64
 import logging
 import pathlib
 import sys
-from typing import Any, AsyncGenerator, Dict, Generator, List, Union
+from typing import Any, AsyncGenerator, Dict, Generator, List, Optional, Union
 
 from fastapi import (
     Depends,
@@ -39,6 +39,16 @@ logger = logging.getLogger(__name__)
 
 core_api = FastAPI()
 
+root_path: Optional[str] = None
+# root_path: Optional[str] = pathlib.Path(sys.executable).anchor
+
+
+@core_api.on_event("startup")
+async def startup_event():
+    global root_path
+    root_path = pathlib.Path(sys.executable).anchor
+    print(f"Event in api.py")
+
 
 def get_runtime_repo() -> Generator[ProjectRepository, None, None]:
     """Fastapi dependencies function creating `repositories` for endpoint."""
@@ -616,17 +626,15 @@ async def get_storage() -> List[Union[Dict[str, Any], str, None]]:
     HTTPException
         If the path is invalid.
     """
-    path = pathlib.Path(sys.executable).anchor
-
     try:
-        return get_directory_listing(path)
+        return get_directory_listing(root_path)
     except NotADirectoryError:
-        logger.warning("Chosen path '{}' is not a directory.".format(path))
+        logger.warning("Chosen path '{}' is not a directory.".format(root_path))
         raise HTTPException(
             status_code=400, detail="Chosen path is not a directory"
         )
     except FileNotFoundError:
-        logger.warning("Chosen path '{}' is not valid.".format(path))
+        logger.warning("Chosen path '{}' is not valid.".format(root_path))
         raise HTTPException(status_code=404, detail="Chosen path is not valid")
 
 
diff --git a/src/core/services.py b/src/core/services.py
index b33657a..d212aa8 100644
--- a/src/core/services.py
+++ b/src/core/services.py
@@ -591,11 +591,11 @@ def get_directory_listing(
     Dict
         jsTree formatted json containing information about root node at path.
     """
-    if path is None:
-        # https://stackoverflow.com/a/65975115/182868
-        # TODO: The root folders should have been set via the new config
-        # system.
-        path = pathlib.Path(sys.executable).anchor
+    # if path is None:
+    #     # https://stackoverflow.com/a/65975115/182868
+    #     # TODO: The root folders should have been set via the new config
+    #     # system.
+    #     path = pathlib.Path(sys.executable).anchor
 
     normalized_path = pathlib.Path(path)
 
diff --git a/tests/unit/core/integration/test_api.py b/tests/unit/core/integration/test_api.py
index dee2781..b0119aa 100644
--- a/tests/unit/core/integration/test_api.py
+++ b/tests/unit/core/integration/test_api.py
@@ -10,6 +10,7 @@ from sqlalchemy.orm import clear_mappers
 from sqlalchemy.orm.session import close_all_sessions
 
 import core
+import core.api
 import core.main
 from core import api, model
 from core.repository import SqlAlchemyProjectRepository as ProjectRepository
@@ -27,6 +28,19 @@ TEST_VIDEO = str(
 )
 
 
+async def test_startup_event():
+    global root_path
+    root_path = "tests/integration/test_data/test_directory_listing_folder"
+    api.root_path = "tests/integration/test_data/test_directory_listing_folder"
+    core.api.root_path = (
+        "tests/integration/test_data/test_directory_listing_folder"
+    )
+    print(f"Event in test_apy.py {api.root_path}")
+
+
+api.core_api.router.on_startup = [test_startup_event]
+
+
 @pytest.fixture
 def setup(tmp_path):
     """Override setup of db for tests."""
@@ -603,6 +617,53 @@ def test_get_job_objects_no_job_project(make_test_data) -> None:
 
 
 def test_get_storage():
+    """Test getting the directory listing from the API."""
+    with TestClient(api.core_api) as client:
+        response = client.get("/storage/")
+        response_data = response.json()
+        print(response_data)
+        assert response.status_code == 200
+        assert response_data[0]["text"] == "test_directory_listing_folder"
+        assert response_data[0]["type"] == "folder"
+        assert "children" in response_data[0]
+
+        children_results = [
+            {
+                "id": "tests/integration/test_data/test_directory_listing_folder/folderA",
+                "text": "folderA",
+                "type": "folder",
+                "children": True,
+            },
+            {
+                "id": "tests/integration/test_data/test_directory_listing_folder/folderB",
+                "text": "folderB",
+                "type": "folder",
+                "children": True,
+            },
+            {
+                "id": "tests/integration/test_data/test_directory_listing_folder/emptyfile",
+                "text": "emptyfile",
+                "type": "file",
+            },
+            {
+                "id": "tests/integration/test_data/test_directory_listing_folder/invalidext.in21p3",
+                "text": "invalidext.in21p3",
+                "type": "file",
+            },
+            {
+                "id": "tests/integration/test_data/test_directory_listing_folder/text.txt",
+                "text": "text.txt",
+                "type": "text/plain",
+            },
+            {
+                "id": "tests/integration/test_data/test_directory_listing_folder/video.mp4",
+                "text": "video.mp4",
+                "type": "video/mp4",
+            },
+        ]
+
+
+def test_get_storage_path():
     """Test getting the directory listing from the API."""
     with TestClient(api.core_api) as client:
         pathb = (
```
